File: frontend/lib/phone.ts
export type NormalizedPhone = {
  local: string;
  e164: string;
  digits: string;
};

const KSA_REGEX_05 = /^05\d{8}$/;
const KSA_REGEX_5 = /^5\d{8}$/;
const KSA_REGEX_966 = /^9665\d{8}$/;
const KSA_REGEX_PLUS966 = /^\+9665\d{8}$/;

export function normalizeKSAPhone(raw: string): NormalizedPhone | null {
  const cleaned = raw.replace(/[\s\-\(\)]/g, "").trim();

  let core: string | null = null;

  if (KSA_REGEX_PLUS966.test(cleaned)) {
    core = cleaned.slice(-9);
  } else if (KSA_REGEX_966.test(cleaned)) {
    core = cleaned.slice(-9);
  } else if (KSA_REGEX_05.test(cleaned)) {
    core = cleaned.slice(1);
  } else if (KSA_REGEX_5.test(cleaned)) {
    core = cleaned;
  }

  if (!core) return null;

  return {
    local: `0${core}`,
    e164: `+966${core}`,
    digits: `966${core}`,
  };
}

export function isValidKSAPhone(raw: string): boolean {
  return normalizeKSAPhone(raw) !== null;
}

File: backend/app/services/phone.py
import re


class PhoneValidationError(ValueError):
    pass


def normalize_ksa_phone(raw: str) -> dict[str, str]:
    """
    Normalize a KSA mobile number to three canonical forms.

    Accepts:
    - 05XXXXXXXX  (10 digits)
    - 5XXXXXXXX   (9 digits)
    - 9665XXXXXXXX (12 digits)
    - +9665XXXXXXXX (13 chars)

    Returns:
    - local:  05XXXXXXXX
    - e164:   +9665XXXXXXXX
    - digits: 9665XXXXXXXX
    """
    cleaned = re.sub(r"[\s\-\(\)]", "", raw.strip())

    # Extract the core 9 digits after 5
    core: str | None = None

    if re.fullmatch(r"\+?9665\d{8}", cleaned):
        core = cleaned[-9:]
    elif re.fullmatch(r"05\d{8}", cleaned):
        core = cleaned[1:]
    elif re.fullmatch(r"5\d{8}", cleaned):
        core = cleaned
    else:
        raise PhoneValidationError(
            "رقم الجوال لازم يكون رقم سعودي صحيح مثل 05XXXXXXXX"
        )

    local = f"0{core}"
    e164 = f"+966{core}"
    digits = f"966{core}"

    return {"local": local, "e164": e164, "digits": digits}
